Tweak windows introduction page (#26848)
kubernetes/website: content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md

Windows {{< glossary_tooltip text="nodes" term_id="node" >}} are
[supported](#windows-os-version-support) provided that the operating system is
Windows Server 2019.

This document uses the term *Windows containers* to mean Windows containers with
process isolation. Kubernetes does not support running Windows containers with
[Hyper-V isolation](https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/hyperv-container).

## Resource management

On Linux nodes, {{< glossary_tooltip text="cgroups" term_id="cgroup" >}} are used
as a pod boundary for resource control. Containers are created within that boundary
for network, process and file system isolation. The Linux cgroup APIs can be used
to gather CPU, I/O, and memory use statistics.

In contrast, Windows uses a _job object_ per container with a system namespace filter
to contain all processes in a container and provide logical isolation from the
host.
(Job objects are a Windows process isolation mechanism and are different from
what Kubernetes refers to as a {{< glossary_tooltip term_id="job" text="Job" >}}).

> There is no way to run a Windows container without the namespace filtering in

Reviewer: We have 'HostProcess' containers now (alpha in v1.22 and hopefully beta in v1.23) which are similar to privilege containers and can access / manipulate host resources.
We can update this section with a follow on PR tho :)

Author: Please do! Or, even better, split that off into a separate page. Remember, this page is _supposed_ to be about setting up a cluster.

Reviewer: We have https://kubernetes.io/docs/tasks/configure-pod-container/create-hostprocess-pod/
I was going to update that info to point to that page :)

(on Linux, the kubelet will by default not start with swap space enabled).

Windows nodes do not overcommit memory for processes running in containers. The
net effect is that Windows won't reach out of memory conditions the same way Linux
does, and processes page to disk instead of being subject to out of memory (OOM)
termination. If memory is over-provisioned and all physical memory is exhausted,
then paging can slow down performance.

You can place bounds on memory use for workloads using the kubelet
parameters `--kubelet-reserve` and/or `--system-reserve`; these account
for memory usage on the node (outside of containers), and reduce
[NodeAllocatable](/docs/tasks/administer-cluster/reserve-compute-resources/#node-allocatable)).
As you deploy workloads, set resource limits on containers. This also subtracts from
`NodeAllocatable` and prevents the scheduler from adding more pods once a node is full.

{{< note >}}
When you set memory resource limits for Windows containers, you should either set a
limit and leave the memory request unspecified, or set the request equal to the limit.
{{< /note >}}

On Windows, good practice to avoid over-provisioning is to configure the kubelet

> with a system reserved memory of at least 2GiB to account for Windows, Kubernetes

Reviewer: ```suggestion
with a system reserved memory of at least 2GiB to account for Windows, Kubernetes,
```

kubelet parameters `--kubelet-reserve` and/or `--system-reserve` to
account for CPU usage on the node (outside of containers).
This reduces `NodeAllocatable`.
The cluster-wide scheduler then takes this reservation into account when determining
pod placement.

On Windows, the kubelet supports a command-line flag to set the priority of the
kubelet process: `--windows-priorityclass`. This flag allows the kubelet process to get
more CPU time slices when compared to other processes running on the Windows host.
More information on the allowable values and their meaning is available at
[Windows Priority Classes](https://docs.microsoft.com/en-us/windows/win32/procthread/scheduling-priorities#priority-class).
To ensure that running Pods do not starve the kubelet of CPU cycles, set this flag to `ABOVE_NORMAL_PRIORITY_CLASS` or above.

## Compatibility and limitations {#limitations}

Some node features are only available if you use a specific
[container runtime](#container-runtime); others are not available on Windows nodes,
including:

* HugePages: not supported for Windows containers
* Privileged containers: not supported for Windows containers

> * TerminationGracePeriod: not implemented

Reviewer: ```suggestion
* TerminationGracePeriod: required containerD
```

subtleties around what's different come down to differences in the OS and container
runtime. In certain situations, some properties on workload resources were designed
under the assumption that they would be implemented on Linux, and fail to run on Windows.

At a high level, these OS concepts are different:

* Identity - Linux uses userID (UID) and groupID (GID) which
  are represented as integer types. User and group names
  are not canonical - they are just an alias in `/etc/groups`
  or `/etc/passwd` back to UID+GID. Windows uses a larger binary
  [security identifier](https://docs.microsoft.com/en-us/windows/security/identity-protection/access-control/security-identifiers) (SID)
  which is stored in the Windows Security Access Manager (SAM) database. This
  database is not shared between the host and containers, or between containers.
* File permissions - Windows uses an access control list based on (SIDs), whereas
  POSIX systems such as Linux use a bitmask based on object permissions and UID+GID,
  plus _optional_ access control lists.
* File paths - the convention on Windows is to use `\` instead of `/`. The Go IO

>   libraries typically accept both and just make it work, but when you're setting a

Reviewer: ```suggestion
  libraries and PowerShell typically accept both and just make it work, but when you're setting a
```

{{< glossary_tooltip text="Pods" term_id="pod" >}} on Kubernetes is similar to
scheduling Linux-based containers.

In order to run Windows containers, your Kubernetes cluster must include
multiple operating systems.
The {{< glossary_tooltip text="control plane" term_id="control-plane" >}} is supported
on Linux only; however, you can deploy nodes running either Windows or Linux depending
on your workload needs.

Windows {{< glossary_tooltip text="nodes" term_id="node" >}} are
[supported](#windows-os-version-support) provided that the operating system is
Windows Server 2019.

This document uses the term “Windows containers” to mean Windows containers with
process isolation. Kubernetes does not support running Windows containers with
[Hyper-V isolation](https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/hyperv-container).

## Resource management

On Linux nodes, {{< glossary_tooltip text="cgroups" term_id="cgroup" >}} are used
as a pod boundary for resource control. Containers are created within that boundary
for network, process and file system isolation. The Linux cgroup APIs can be used
to gather CPU, I/O, and memory use statistics.

> In contrast, Windows uses a _job object_ per container with a system namespace filter

Author: I'm not sure this is right. Is the “job object” obviously within the container?

Containers cannot assume an identity from the host because the Security Account Manager
(SAM) is separate.

#### Memory reservations {#resource-management-memory}

Windows does not have an out-of-memory process killer as Linux does. Windows always
treats all user-mode memory allocations as virtual, and pagefiles are mandatory
(on Linux, the kubelet will by default not start with swap space enabled).

Windows nodes do not overcommit memory for processes running in containers. The
net effect is that Windows won't reach out of memory conditions the same way Linux
does, and processes page to disk instead of being subject to out of memory (OOM)
termination. If memory is over-provisioned and all physical memory is exhausted,
then paging can slow down performance.

You can place bounds on memory use for workloads using the kubelet
parameters `--kubelet-reserve` and/or `--system-reserve`; these account
for memory usage on the node (outside of containers), and reduce
[NodeAllocatable](/docs/tasks/administer-cluster/reserve-compute-resources/#node-allocatable)).
As you deploy workloads, set resource limits on containers. This also subtracts from

> `NodeAllocatable` and prevents the scheduler from adding more pods once a node is full.

Author: We typically wouldn't write pods in title case here. Other localizations might, especially where “pods” is not a word in the target language.

Kubernetes host processes such as the kubelet, you can (and should) reserve a
percentage of total CPU. You should determine this CPU reservation taking account of
to the number of CPU cores available on the node. To decide on the CPU percentage to
reserve, identify the maximum pod density for each node and monitor the CPU usage of
the system services running there, then choose a value that meets your workload needs.

You can place bounds on CPU usage for workloads using the
kubelet parameters `--kubelet-reserve` and/or `--system-reserve` to
account for CPU usage on the node (outside of containers).
This reduces `NodeAllocatable`.
The cluster-wide scheduler then takes this reservation into account when determining
pod placement.

On Windows, the kubelet supports a command-line flag to set the priority of the
kubelet process: `--windows-priorityclass`. This flag allows kubelet process to get
more CPU time slices when compared to other processes running on the Windows host.
More information on the allowable values and their meaning is available at
[Windows Priority Classes](https://docs.microsoft.com/en-us/windows/win32/procthread/scheduling-priorities#priority-class).

> To ensure that running Pods do not starve the kubelet of CPU cycles, set this flag to `ABOVE_NORMAL_PRIORITY_CLASS` or above.

Reviewer 2: these are constants defined here: https://docs.microsoft.com/en-us/windows/win32/procthread/scheduling-priorities#priority-class (link is reference in the paragraph above.  You have to choose one of these constants otherwise kubelet fails to start: https://github.com/kubernetes/kubernetes/blob/c1e69551be1a72f0f8db6778f20658199d3a686d/cmd/kubelet/app/init_windows.go#L56-L60

  Linux containers in the same Pod. All containers in a Pod are scheduled onto a single
  Node where each Node represents a specific platform and architecture. The following
  Pod capabilities, properties and events are supported with Windows containers:

  * Single or multiple containers per Pod with process isolation and volume sharing
  * Pod `status` fields
  * Readiness and Liveness probes
  * postStart & preStop container lifecycle events
  * ConfigMap, Secrets: as environment variables or volumes
  * `emptyDir` volumes
  * Named pipe host mounts
  * Resource limits
* [Workload resources](/docs/concepts/workloads/controllers/) including:
  * ReplicaSet
  * Deployments
  * StatefulSets
  * DaemonSet
  * Job
  * CronJob
  * ReplicationController
* {{< glossary_tooltip text="Services" term_id="service" >}}\
  See [Load balancing and Services](#load-balancing-and-services) for more details.

Pods, workload resources, and Services are critical elements to managing Windows
workloads on Kubernetes. However, on their own they are not enough to enable
the proper lifecycle management of Windows workloads in a dynamic cloud native

> environment. Kubernetes also supports:

Author: I'm not sure what the original text meant to imply!

> environment. Kubernetes also supports:

Reviewer: I think maybe these functionalities were not available when Windows container support went stable (v1.14) but has since been added?

I wonder if we should just remove parts of this section because I am also not sure what the original text meant to imply.

with the [Troubleshooting](/docs/tasks/debug-application-cluster/troubleshooting/)
task page.

Some additional, Windows-specific troubleshooting help is included
in this section. Logs are an important element of troubleshooting
issues in Kubernetes. Make sure to include them any time you seek
troubleshooting assistance from other contributors. Follow the
instructions in the
SIG Windows [contributing guide on gathering logs](https://github.com/kubernetes/community/blob/master/sig-windows/CONTRIBUTING.md#gathering-logs).

### Node-level troubleshooting {#troubleshooting-node}

1. How do I know `start.ps1` completed successfully?

   You should see kubelet, kube-proxy, and (if you chose Flannel as your networking
   solution) flanneld host-agent processes running on your node, with running logs
   being displayed in separate PowerShell windows. In addition to this, your Windows
   node should be listed as "Ready" in your Kubernetes cluster.

1. Can I configure the Kubernetes node processes to run in the background as services?

   The kubelet and kube-proxy are already configured to run as native Windows Services,

>    offering resiliency by re-starting the services automatically in the event of

Author: These are not Services in the Kubernetes API. “services” is less likely to mislead.

   failure (for example a process crash). You have two options for configuring these

>    node components as services.

Author: These are not Services in the Kubernetes API. “services” is less likely to mislead.

{{< note >}}
The Kubernetes control plane, including the [master components](/docs/concepts/overview/components/), continues to run on Linux. There are no plans to have a Windows-only Kubernetes cluster.
{{< /note >}}

{{< note >}}
In this document, when we talk about Windows containers we mean Windows containers with process isolation. Windows containers with [Hyper-V isolation](https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/hyperv-container) is planned for a future release.
{{< /note >}}

## Supported Functionality and Limitations

### Supported Functionality

#### Windows OS Version Support

Refer to the following table for Windows operating system support in Kubernetes. A single heterogeneous Kubernetes cluster can have both Windows and Linux worker nodes. Windows containers have to be scheduled on Windows nodes and Linux containers on Linux nodes.

> | Kubernetes version | Windows Server LTSC releases | Windows Server SAC releases |

Reviewer: I think we should keep the Windows OS Version Support section in there.
Users frequently ask which Windows versions have been tested to work with Kubernetes and also there is occasionally functionality added to Kubernetes that depends on OS support.

Author: This is a long page for an _introduction_ - could I move the support table to the reference section and hyperlink there?

Reviewer: That should work.

> This document uses the term “Windows containers” to mean Windows containers with process isolation. Kubernetes does not support running Windows containers with [Hyper-V isolation](https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/hyperv-container).

Author: That might be a separate update. If the existing document is incorrect we could fix that separately then I can rebase.

I don't know if containerd and Hyper-V isolation works; it's well outside my area of technical expertise.

> This document uses the term “Windows containers” to mean Windows containers with process isolation. Kubernetes does not support running Windows containers with [Hyper-V isolation](https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/hyperv-container).

Reviewer 2: containerd with process isolation today works.  Hyper-v will be possible but is not currently up streamed in containerd and not supported in k8s.  you can try it out really early builds at https://github.com/Azure/aks-engine/blob/96c2dd4d91f2cc7fa807da7c83170267404044a7/docs/topics/features.md#hyper-v-support

> on the supported Windows versions.

Reviewer: Or possibly `validated`?
We do run e2e tests that cover everything in that matrix.

- jsturtevant
- marosset
- perithompson
title: Windows containers in Kubernetes
content_type: concept
weight: 20
---

<!-- overview -->

Windows applications constitute a large portion of the services and applications that run in many organizations. [Windows containers](https://aka.ms/windowscontainers) provide a way to encapsulate processes and package dependencies, making it easier to use DevOps practices and follow cloud native patterns for Windows applications.

Organizations with investments in Windows-based applications and Linux-based applications don't have to look for separate orchestrators to manage their workloads, leading to increased operational efficiencies across their deployments, regardless of operating system.

<!-- body -->

## Windows nodes in Kubernetes

To enable the orchestration of Windows containers in Kubernetes, include Windows nodes in your existing Linux cluster. Scheduling Windows containers in {{< glossary_tooltip text="Pods" term_id="pod" >}} on Kubernetes is similar to scheduling Linux-based containers.

> In order to run Windows containers, your Kubernetes cluster must include multiple operating systems. The {{< glossary_tooltip text="control plane" term_id="control-plane" >}} is supported on Linux only; however, you can deploy nodes running either Windows or Linux depending on your workload needs. Windows {{< glossary_tooltip text="nodes" term_id="node" >}} are i[supported](#windows-os-version-support) provided that the operating system is Windows Server 2019.

Reviewer 3: nit: remove extra "i"

Author: Thanks!

Learn how to [install ContainerD on a Windows](/docs/setup/production-environment/container-runtimes/#install-containerd).

{{< note >}}
There is a [known limitation](/docs/tasks/configure-pod-container/configure-gmsa/#gmsa-limitations) when using GMSA with ContainerD to access Windows network shares which requires a kernel patch.
{{< /note >}}

#### Pause image

Microsoft maintains a Windows pause infrastructure container at `mcr.microsoft.com/oss/kubernetes/pause:1.4.1`.

> ### Container runtimes {#container-runtime}

Reviewer: I think this ### Container runtimes may be a duplicate and not needed.

For the node, pod, and service objects, the following network flows are supported for TCP/UDP traffic:

* Pod → Pod (IP)
* Pod → Pod (Name)
* Pod → Service (Cluster IP)
* Pod → Service (PQDN, but only if there are no ".")
* Pod → Service (FQDN)
* Pod → external (IP)
* Pod → external (DNS)
* Node → Pod
* Pod → Node

#### CNI plugin limitations

* Windows reference network plugins win-bridge and win-overlay do not implement [CNI spec](https://github.com/containernetworking/cni/blob/master/SPEC.md) v0.4.0, due to a missing "CHECK" implementation.

> * The Flannel VXLAN CNI plugin has the following limitations on Windows:

Reviewer 3: @sftim, Would you double check this list, starting at line 170?
See ,https://deploy-preview-26848--kubernetes-io-master-staging.netlify.app/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#cni-plugin-limitations

* Eviction by using `--enforce-node-allocable` is not implemented
* Eviction by using `--eviction-hard` and `--eviction-soft` are not implemented
* The `MemoryPressure` Condition is not implemented
* There are no OOM eviction actions taken by the kubelet
* A kubelet running on a Windows node does not have memory restrictions. `--kubelet-reserve` and `--system-reserve` do not set limits on kubelet or processes running on the host. This means kubelet or a process on the host could cause memory resource starvation outside the node-allocatable and scheduler
* An additional flag to set the priority of the kubelet process is available on the Windows nodes called `--windows-priorityclass`. This flag allows kubelet process to get more CPU time slices when compared to other processes running on the Windows host. More information on the allowable values and their meaning is available at [Windows Priority Classes](https://docs.microsoft.com/en-us/windows/win32/procthread/scheduling-priorities#priority-class). In order for the kubelet to always have enough CPU cycles it is recommended to set this flag to `ABOVE_NORMAL_PRIORITY_CLASS` and above

### Operating system restrictions

> Windows has strict compatibility rules, where the host OS version must match the container base image OS version. Only Windows containers with a container operating system of Windows Server 2019 are supported.

Reviewer 4: Should we link to https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-20H2%2Cwindows-10-20H2 somewhere here as it might help the user?

Author: That might be a separate update - this one's already large.

### Node-level troubleshooting {#troubleshooting-node}

1. How do I know `start.ps1` completed successfully?

   You should see kubelet, kube-proxy, and (if you chose Flannel as your networking solution) flanneld host-agent processes running on your node, with running logs being displayed in separate PowerShell windows. In addition to this, your Windows node should be listed as "Ready" in your Kubernetes cluster.

1. Can I configure the Kubernetes node processes to run in the background as services?

   The kubelet and kube-proxy are already configured to run as native Windows Services, offering resiliency by re-starting the services automatically in the event of failure (for example a process crash). You have two options for configuring these node components as services.

    1. As native Windows Services

>         You can run the kubelet & kube-proxy as native Windows Services using `sc.exe`.

Reviewer 3: nit: s/&/and/

Author: ```suggestion
        You can run the kubelet and kube-proxy as native Windows Services using `sc.exe`.
```

{{< /table >}}

##### Session affinity

Setting the maximum session sticky time for Windows services using `service.spec.sessionAffinityConfig.clientIP.timeoutSeconds` is not supported.

#### DNS {#dns-limitations}

* ClusterFirstWithHostNet is not supported for DNS. Windows treats all names with a `.` as a FQDN and skips FQDN resolution
* On Linux, you have a DNS suffix list, which is used when trying to resolve PQDNs. On Windows, you can only have 1 DNS suffix, which is the DNS suffix associated with that pod's namespace (mydns.svc.cluster.local for example). Windows can resolve FQDNs and services or names resolvable with just that suffix. For example, a pod spawned in the default namespace, will have the DNS suffix **default.svc.cluster.local**.
  Inside a Windows pod, you can resolve both **kubernetes.default.svc.cluster.local** and **kubernetes**, but not the in-betweens, like **kubernetes.default** or **kubernetes.default.svc**.
* On Windows, there are multiple DNS resolvers that can be used. As these come with slightly different behaviors, using the `Resolve-DNSName` utility for name query resolutions is recommended.

#### IPv4/IPv6 dual-stack

> Kubernetes on Windows does not support single-stack "IPv6-only" networking. However,dual-stack IPv4/IPv6 networking for pods and nodes with single-family services is supported.

Reviewer 2: @daschott is single stack IPv6 supported in 2004+?

* `V1.Container.SecurityContext.allowPrivilegeEscalation` - not possible on Windows, none of the capabilities are hooked up
* `V1.Container.SecurityContext.Capabilities` - POSIX capabilities are not implemented on Windows
* `V1.Container.SecurityContext.privileged` - Windows doesn't support privileged containers
* `V1.Container.SecurityContext.procMount` - Windows doesn't have a `/proc` filesystem
* `V1.Container.SecurityContext.readOnlyRootFilesystem` - not possible on Windows, write access is required for registry & system processes to run inside the container
* `V1.Container.SecurityContext.runAsGroup` - not possible on Windows, no GID support
* `V1.Container.SecurityContext.runAsNonRoot` - Windows does not have a root user. The closest equivalent is ContainerAdministrator which is an identity that doesn't exist on the node.
* `V1.Container.SecurityContext.runAsUser` - not possible on Windows, no UID support as int.
* `V1.Container.SecurityContext.seLinuxOptions` - not possible on Windows, no SELinux

> * `V1.Container.terminationMessagePath` - this has some limitations in that Windows doesn't support mapping single files. The default value is `/dev/termination-log`, which does work because it does not exist on Windows by default.

Reviewer 2: since we now have single file mapping, I believe this should be available in containerd deployments 1.21+

Author: Can you suggest the replacement text @jsturtevant? I don't think I can work out what this should say instead.

Reviewer 2: @claudiubelu do you know if this `termination-log` works now with single file mapping containerd?

Reviewer 5: Yep, it does. We've also added support for this in Kubernetes v1.17: https://github.com/kubernetes/kubernetes/pull/83057

##### Field compatibility for Pod specifications {#compatibility-v1-pod}

The following list documents differences between how Pod specifications work between Windows and Linux:

* `V1.Pod.hostIPC`, `v1.pod.hostpid` - host namespace sharing is not possible on Windows
* `V1.Pod.hostNetwork` - There is no Windows OS support to share the host network
* `V1.Pod.dnsPolicy` - ClusterFirstWithHostNet - is not supported because Host Networking is not supported on Windows.
* `V1.Pod.podSecurityContext` - see V1.PodSecurityContext below
* `V1.Pod.shareProcessNamespace` - this is a beta feature, and depends on Linux namespaces which are not implemented on Windows. Windows cannot share process namespaces or the container's root filesystem. Only the network can be shared.

> * `V1.Pod.terminationGracePeriodSeconds` - this is not fully implemented in Docker on Windows, see: [reference](https://github.com/moby/moby/issues/25982). The behavior today is that the ENTRYPOINT process is sent CTRL_SHUTDOWN_EVENT, then Windows waits 5 seconds by default, and finally shuts down all processes using the normal Windows shutdown behavior. The 5 second default is actually in the Windows registry [inside the container](https://github.com/moby/moby/issues/25982#issuecomment-426441183), so it can be overridden when the container is built.

Reviewer 2: To be clear we may want to say that it is supported in containerd.

Author: So `terminationGracePeriodSeconds` works with containerd fully, and partially if your container runtime is Docker. Have I got that right?

Reviewer 2: yes

* `V1.PodSecurityContext.RunAsNonRoot` - Windows does not have a root user. The closest equivalent is ContainerAdministrator which is an identity that doesn't exist on the node.
* `V1.PodSecurityContext.SupplementalGroups` - provides GID, not available on Windows
* `V1.PodSecurityContext.Sysctls` - these are part of the Linux sysctl interface. There's no equivalent on Windows.

### Node problem detector

The node problem detector (see [Monitor Node Health](/docs/tasks/debug-application-cluster/monitor-node-health/)) is not compatible with Windows.

### Pause container

In a Kubernetes Pod, an infrastructure or “pause” container is first created to host the container. In Linux, the cgroups and namespaces that make up a pod need a process to maintain their continued existence; the pause process provides this. Containers that belong to the same pod, including infrastructure and worker containers, share a common network endpoint (same IPv4 and / or IPv6 address, same network port spaces). Kubernetes uses pause containers to allow for worker containers crashing or restarting without losing any of the networking configuration.

> Microsoft maintains a Windows pause infrastructure container at `mcr.microsoft.com/oss/kubernetes/pause:1.4.1`. For more details, see the [DOCKERFILE](https://github.com/kubernetes-sigs/windows-testing/blob/master/images/pause/Dockerfile).

Reviewer 2: ```suggestion
Microsoft maintains a Windows pause infrastructure container at `mcr.microsoft.com/oss/kubernetes/pause:3.4.1`. For more details, see the [DOCKERFILE](https://github.com/kubernetes-sigs/windows-testing/blob/master/images/pause/Dockerfile).
```

Reviewer 2: The pause image is now built along side the same versioning as Linux

Author: Is it reasonable to remove any mention of a `pause` container image from this page?

Reviewer 2: I think we should keep it for a release or two since it was a concern to folks prior to this update

### Container runtimes {#container-runtime}

You need to install a
{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}
into each node in the cluster so that Pods can run there.

These container runtimes work with Windows.

{{% thirdparty-content %}}

#### cri-containerd

{{< feature-state for_k8s_version="v1.20" state="stable" >}}

You can use {{< glossary_tooltip term_id="containerd" text="ContainerD" >}} 1.4.0+ as the container runtime for Kubernetes nodes that run Windows.

Learn how to [install ContainerD on a Windows](/docs/setup/production-environment/container-runtimes/#install-containerd).

{{< note >}}

> There is a [known limitation](/docs/tasks/configure-pod-container/configure-gmsa/#gmsa-limitations) when using GMSA with containerd to access Windows network shares, which requires a kernel patch.

Reviewer 2: This has been resolved with patch March 2021+(KB5000822) with the following setting:

```
reg add "HKLM\SYSTEM\CurrentControlSet\Services\hns\State" /v EnableCompartmentNamespace /t REG_DWORD /d 1.
```

ref https://github.com/microsoft/Windows-Containers/issues/44#issuecomment-802404294

[Docker EE](https://docs.mirantis.com/containers/v3.0/dockeree-products/dee-intro.html)-basic 19.03+ is available as a container runtime for all Windows Server versions. This works with the legacy dockershim adapter.

See [Install Docker](https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/deploy-containers-on-server#install-docker) for more information.

## Windows OS version compatibility {#windows-os-version-support}

Windows has strict compatibility rules, where the host OS version must match the container base image OS version. Only Windows containers with a container operating system of Windows Server 2019 are supported.

The following table shows Windows operating system compatibility with Kubernetes:

| Kubernetes version | Windows Server LTSC releases | Windows Server SAC releases |
| --- | --- | --- |
| *Kubernetes v1.17* | Windows Server 2019 | Windows Server ver 1809 |
| *Kubernetes v1.18* | Windows Server 2019 | Windows Server ver 1809, Windows Server ver 1903, Windows Server ver 1909 |
| *Kubernetes v1.19* | Windows Server 2019 | Windows Server ver 1909, Windows Server ver 2004 |

> | *Kubernetes v1.20* | Windows Server 2019 | Windows Server ver 1909, Windows Server ver 2004 |

Reviewer 2: Is this maintained somewhere else? I believe we recently updated this list for v1.21

Author: I might have missed that in a rebase - let me check.

Author: If there's documentation we can link to, we can use that. I did previously consider moving this list of supported versions into the Reference section of the docs, and then linking there.

Reviewer 2: the updates for this were merged yesterday: https://github.com/kubernetes/website/pull/26486

The Kubernetes [version-skew policy](/docs/setup/release/version-skew-policy/) also applies.

## Security for Windows nodes {#security}

On Windows, data from Secrets are written out in clear text onto the node's local storage (as compared to using tmpfs / in-memory filesystems on Linux). As a cluster operator, you should take both of the following additional measures:

1. Use file ACLs to secure the Secrets' file location.
1. Apply volume-level encryption using [BitLocker](https://docs.microsoft.com/en-us/windows/security/information-protection/bitlocker/bitlocker-how-to-deploy-on-windows-server).

> [RunAsUsername](/docs/tasks/configure-pod-container/configure-runasusername) can be specified for Windows Pods or containers to execute the container processes as a node-default user. This is roughly equivalent to [RunAsUser](/docs/concepts/policy/pod-security-policy/#users-and-groups).

Reviewer 2: We should add that the recommendation would be to run pods with `ContainerUser`.

Reviewer 2: or create there own user that is not `ContainerAdministrator`

Author: I'm not able to infer what to recommend - it's been a few years since I've used Windows, and not at all with Kubernetes, so I don't feel able to explain what readers should do. @jsturtevant would you be willing to suggest the new wording?

Reviewer 2: We could add a sentence something like: 

> There are two default accounts for Windows containers: `ContainerAdministrator` and `ContainerUser`.  It is recommended to use `ContainerUser` for most workloads, see more details at [Microsoft documentation on container user accounts](https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/container-security#when-to-use-containeradmin-and-containeruser-user-accounts)

  * ICMP packets directed to pass through a remote network (e.g. pod to external internet communication via ping) cannot be transposed and thus will not be routed back to their source
  * Since TCP/UDP packets can still be transposed, one can substitute `ping <destination>` with `curl <destination>` to be able to debug connectivity to the outside world.

#### CNI plugins

* Windows reference network plugins win-bridge and win-overlay do not implement [CNI spec](https://github.com/containernetworking/cni/blob/master/SPEC.md) v0.4.0, due to a missing "CHECK" implementation.
* The Flannel VXLAN CNI plugin has the following limitations on Windows:

1. Node-pod connectivity isn't possible by design. It's only possible for local pods with Flannel v0.12.0 (or higher).

> 2. Flannel is restricted to using VNI 4096 and UDP port 4789. See the official [Flannel VXLAN](https://github.com/coreos/flannel/blob/master/Documentation/backends.md#vxlan) backend docs for more details on these parameters.

Reviewer 4: Might be useful to call out https://docs.microsoft.com/en-us/virtualization/windowscontainers/kubernetes/common-problems#pod-to-pod-connectivity-between-hosts-is-broken-on-my-kubernetes-cluster-running-on-vsphere in the context of port 4789.

Author: That might be a separate update - this one's already large.

       nssm.exe set kube-proxy AppEnvironmentExtra KUBE_NETWORK=cbr0
       nssm set kube-proxy DependOnService kubelet
       nssm start kube-proxy

       # Register kube-proxy.exe (overlay / vxlan)
       nssm install kube-proxy C:\k\kube-proxy.exe
       nssm set kube-proxy AppDirectory c:\k
       nssm set kube-proxy AppParameters --v=4 --proxy-mode=kernelspace --feature-gates="WinOverlay=true" --hostname-override=<hostname> --kubeconfig=c:\k\config --network-name=vxlan0 --source-vip=<source-vip> --enable-dsr=false --log-dir=<log directory> --logtostderr=false
       nssm set kube-proxy DependOnService kubelet
       nssm start kube-proxy
       ```

       For initial troubleshooting, you can use the following flags in [nssm.exe](https://nssm.cc/) to redirect stdout and stderr to a output file:

       ```powershell
       nssm set <Service Name> AppStdout C:\k\mysvc.log
       nssm set <Service Name> AppStderr C:\k\mysvc.log
       ```

       For additional details, see [NSSM - the Non-Sucking Service Manager](https://nssm.cc/usage).

1. My Pods are stuck at "Container Creating" or restarting over and over

>    Check that your pause image is compatible with your OS version. The [instructions](https://docs.microsoft.com/en-us/virtualization/windowscontainers/kubernetes/deploying-resources) assume that both the OS and the containers are version 1803. If you have a later version of Windows, such as an Insider build, you need to adjust the images accordingly.

Reviewer 2: ```suggestion
   Check that your pause image is compatible with your OS version. The [instructions](https://docs.microsoft.com/en-us/virtualization/windowscontainers/kubernetes/deploying-resources) assume that both the OS and the containers are the same version (i.e 2019 LTSC). If you have a later version of Windows, such as an Insider build, you need to adjust the images accordingly.
```

   ```

1. My Windows node cannot access my services using the service IP

   This is a known limitation of the networking stack on Windows. Windows Pods are able to access the service IP however.

1. No network adapter is found when starting the kubelet

   The Windows networking stack needs a virtual adapter for Kubernetes networking to work. If the following commands return no results (in an admin shell), virtual network creation — a necessary prerequisite for Kubelet to work — has failed:

   ```powershell
   Get-HnsNetwork | ? Name -ieq "cbr0"
   Get-NetAdapter | ? Name -Like "vEthernet (Ethernet*"
   ```

   Often it is worthwhile to modify the [InterfaceName](https://github.com/microsoft/SDN/blob/master/Kubernetes/flannel/start.ps1#L7) parameter of the start.ps1 script, in cases where the host's network adapter isn't "Ethernet". Otherwise, consult the output of the `start-kubelet.ps1` script to see if there are errors during virtual network creation.

1. DNS resolution is not properly working

   Check the DNS limitations for Windows in this [section](#dns-limitations).

1. `kubectl port-forward` fails with "unable to do port forwarding: wincat not found"

>    This was implemented in Kubernetes 1.15 by including `wincat.exe` in the pause infrastructure container `mcr.microsoft.com/oss/kubernetes/pause:1.4.1`. Be sure to use a supported version of Kubernetes.

Reviewer 2: ```suggestion
   This was implemented in Kubernetes 1.15 by including `wincat.exe` in the pause infrastructure container `mcr.microsoft.com/oss/kubernetes/pause:3.4.1`. Be sure to use a supported version of Kubernetes.
```

Author: The original page was incorrect about the version that included the fix?

>    If you would like to build your own pause infrastructure container be sure to include [wincat](https://github.com/kubernetes-sigs/sig-windows-tools/tree/master/cmd/wincat).

Reviewer 2: ```suggestion
   If you would like to build your own pause infrastructure container be sure to include [wincat](https://github.com/kubernetes/kubernetes/tree/master/build/pause/windows/wincat).
```

* [Relevant logs](https://github.com/kubernetes/community/blob/master/sig-windows/CONTRIBUTING.md#gathering-logs)
* Tag the issue sig/windows by commenting on the issue with `/sig windows` to bring it to a SIG-Windows member's attention

## {{% heading "whatsnext" %}}

### Deployment tools

The kubeadm tool helps you to deploy a Kubernetes cluster, providing the control plane to manage the cluster it, and nodes to run your workloads. [Adding Windows nodes](/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes/) explains how to deploy Windows nodes to your cluster using kubeadm.

The Kubernetes [cluster API](https://cluster-api.sigs.k8s.io/) project also provides means to automate deployment of Windows nodes.

### Windows distribution channels

For a detailed explanation of Windows distribution channels see the [Microsoft documentation](https://docs.microsoft.com/en-us/windows-server/get-started-19/servicing-channels-19).

> Information on the different Windows Server servicing channels including their support models can be found at [Windows Server servicing channels](https://docs.microsoft.com/en-us/windows-server/get-started-19/servicing-channels-19).

Reviewer 2: This line and the distribution channels line right above it link to the same page.  I believe "distribution channels" and "servicing channels" mean the same thing in this context and only need to be referenced once.  @immuzz could you confirm?
